services/FeedbackService.ts

```diff
@@ -0,0 +1,66 @@
+import { Service } from 'typedi';
+import { EntityManager } from 'typeorm';
+import { InjectManager } from 'typeorm-typedi-extensions';
+import * as moment from 'moment';
+import { FeedbackModel } from '../models/FeedbackModel';
+import { UserModel } from '../models/UserModel';
+import Repositories, { TransactionsManager } from '../repositories';
+import { PublicFeedback, Feedback, Uuid, ActivityType } from '../types';
+import { UserError } from '../utils/Errors';
+import { Config } from '../config';
+
+@Service()
+export default class FeedbackService {
+  private transactions: TransactionsManager;
+
+  constructor(@InjectManager() entityManager: EntityManager) {
+    this.transactions = new TransactionsManager(entityManager);
+  }
+
+  public async getFeedback(canSeeAllFeedback = false, user: UserModel): Promise<PublicFeedback[]> {
+    const feedback = await this.transactions.readOnly(async (txn) => {
+      const feedbackRepository = Repositories.feedback(txn);
+      if (canSeeAllFeedback) return feedbackRepository.getFeedback();
+      return feedbackRepository.getFeedbackByUser(user);
+    });
+    return feedback.map((fb) => fb.getPublicFeedback());
+  }
+
+  public async submitFeedback(user: UserModel, feedback: Feedback): Promise<PublicFeedback> {
+    const feedbackObject = FeedbackModel.create({ ...feedback, user });
+    const addedFeedback = await this.transactions.readWrite(async (txn) => {
+      const feedbackRepository = Repositories.feedback(txn);
+
+      const latestFeedback = await feedbackRepository.getLatestFeedback(user);
+      if (latestFeedback) {
+        const oneWeekAfterFeedback = moment(latestFeedback.timestamp).add(1, 'week');
+        if (oneWeekAfterFeedback.valueOf() > moment.now()) {
+          const daysLeft = oneWeekAfterFeedback.diff(moment(), 'days', false) + 1;
+          throw new UserError(`You must wait ${daysLeft} days to submit more feedback`);
+        }
+      }
+
+      const fb = await feedbackRepository.upsertFeedback(feedbackObject);
+      return fb;
+    });
+    return addedFeedback.getPublicFeedback();
+  }
+
+  public async acknowledgeFeedback(uuid: Uuid) {
+    const acknowledgedFeedback = await this.transactions.readWrite(async (txn) => {
+      const feedbackRepository = Repositories.feedback(txn);
+      const feedback = await feedbackRepository.findByUuid(uuid);
+      if (!feedback) throw new UserError('Feedback not found');
+      if (feedback.acknowledged) throw new UserError('This feedback has already been acknowledged');
```

Comment:
But if this route is reached with a feedback that has already been acknowledged, doesn't that indicate some kind of user error made by the API caller? It might be confusing to return a successful response when nothing actually happened and that request shouldn't have been made.

Reply:
If you think of this request as "Acknowledge this unacknowledged feedback" then yes it's a user error. If you think of it as "Make sure this feedback is marked as acknowledged" then it's fine to return early. I prefer the latter framing because if two people acknowledge it at the same time, who cares which request goes through? To both of them, they marked it as acknowledged and don't care what its status was before.